src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -473,176 +574,223 @@ void LFUDAPolicy::cleaning(const DoutPrefixProvider* dpp)
-        block.size = cur_len;
-        block.blockID = fst;
-        op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
-        if (op_ret < 0) {
-	  ldpp_dout(dpp, 0) << __func__ << "updating dirty flag in block directory failed, ret=" << op_ret << dendl;
-        }
-        fst += cur_len;
-      } while(fst < lst);
-
-      cacheDriver->rename(dpp, head_oid_in_cache, new_head_oid_in_cache, null_yield);
-      //data is clean now, updating in-memory metadata for an object
-      e->dirty = false;
-      //invoke update() with dirty flag set to false, to update in-memory metadata for head
-      this->update(dpp, new_head_oid_in_cache, 0, 0, e->version, false, y);
-
-      rgw::d4n::CacheBlock block;
-      block.cacheObj.bucketName = c_obj->get_bucket()->get_bucket_id();
-      block.cacheObj.objName = c_obj->get_name();
-      block.size = 0;
-      block.blockID = 0;
-      if (c_obj->have_instance()) {
-        blockDir->get(dpp, &block, null_yield);
-        if (block.version == c_obj->get_instance()) { //versioned case - update head block entry that has latest version
-          op_ret = blockDir->update_field(dpp, &block, "dirty", "false", null_yield);
-          if (op_ret < 0) {
-              ldpp_dout(dpp, 20) << __func__ << "updating dirty flag in block directory for head failed!" << dendl;
+	std::unique_ptr<rgw::sal::Object> c_obj = c_bucket->get_object(e->obj_key);
+	ldpp_dout(dpp, 20) << __func__ << "(): c_obj oid =" << c_obj->get_oid() << dendl;
+
+	std::string prefix = url_encode(e->bucket_id) + CACHE_DELIM + url_encode(e->version) + CACHE_DELIM + url_encode(c_obj->get_name());
+	std::string head_oid_in_cache = DIRTY_BLOCK_PREFIX + prefix;
+
+	// If the state is invalid, the blocks must be deleted from the cache rather than written to the backend.
+	if (invalid) {
+	  ldpp_dout(dpp, 10) << __func__ << "(): State is INVALID; deleting object." << dendl;
+	  int ret = -1;
+	  if ((ret = cacheDriver->delete_data(dpp, head_oid_in_cache, y)) == 0) { // Sam: do we want del or delete_data here?
+	    if (!(ret = erase(dpp, prefix, y))) {
+	      ldpp_dout(dpp, 0) << "Failed to delete head policy entry for: " << head_oid_in_cache << ", ret=" << ret << dendl;
+	      continue;
+	    }
+	  } else {
+	    ldpp_dout(dpp, 0) << "Failed to delete head object for: " << head_oid_in_cache << ", ret=" << ret << dendl;
+	    continue;
+	  }
+
+	  ret = delete_data_blocks(dpp, c_obj->get_name(), e, y);
+          if (ret == 0) {
+	    erase_dirty_object(dpp, e->key, null_yield);
+          } else {
+	    ldpp_dout(dpp, 0) << "Failed to delete blocks for: " << head_oid_in_cache << ", ret=" << ret << dendl;
+	    continue;
           }
+	} else {
+	  const auto diff = std::chrono::system_clock::now() - std::chrono::system_clock::from_time_t(e->creationTime); 
+          if (std::chrono::round<std::chrono::milliseconds>(diff) > std::chrono::milliseconds(interval)) {
```

Comment:
why is this change needed? I had intentionally put the thread to sleep for milliseconds. Did you face a problem here?

Reply:
This is effectively the same as what is being done in the code currently - you have just converted the difference to milliseconds and then compared with `interval`, but the value of diff and interval still are the same.

Reply:
I wanted to make sure the interval was being read in milliseconds because I wasn't sure if the default unit was seconds, and it seems like the cache cleaning method was being put to sleep for a unit value that differed from the interval. It was more observational but I thought that converting the interval to milliseconds this way made the unit clear and the code more readable. I can undo this change if you feel it is unnecessary, though

Reply:
I had intentionally converted the unit to milliseconds while making the thread sleep else the thread would sleep for a long time and the expired elements would pile up, so for now please revert your changes